### ISSUE TITLE
Referee 2.3

### ISSUE THREAD

**Issue by cgreene** [greenelab/deep-review#690]
> The main issue with machine learning solutions in a medical, particularly clinical or public health setting is the lack of proper measures of uncertainty, as it is traditionally provided either in the framework of hypothesis testing or in the increasing acceptance of posterior Bayesian inference for public health decisions. Although this is mentioned throughout the review, this issue deserves a much more prominent place in the introduction and the discusssion, since it is one of the key obstacles for the acceptance of machine learning approaches outside exploratory analyses in basic biological research.

**Comment by cgreene** [comment]
Does anyone feel comfortable addressing this? We can talk a bit about uncertainty/stability but I don't know of good examples in the deep learning literature.

**Comment by cgreene** [comment]
Tagging @akundaje in the hopes that he has some material readily at hand.

**Comment by akundaje** [comment]
Yes. There is a decent amount of literature on this in deep learning and a few applications in the medical imaging space, nothing in genomics or drug discovery AFAIK. @AvantiShri has been working on this in my lab. Av - could likely contribute a paragraph or so with references on uncertainty.

**Comment by akundaje** [comment]
This is a good recent blog post on the topic with relevant references and papers https://alexgkendall.com/computer_vision/bayesian_deep_learning_for_safe_ai/

**Comment by cgreene** [comment]
@AvantiShri - do you want to contribute this in the next week?

**Comment by AvantiShri** [comment]
Next week is a bit busy for me but I could with the help of Amr Alexandari, a master's student I am working with (github id is @alexandari - for some reason I can't tag him?).

I'd like to highlight some points regarding how Yarin Gal's work applies to classification problems. When it comes to classification, Gal himself wrote in that blog post Anshul linked to: "It is important to note that the model uncertainty in the softmax output can be summarised by taking the mean of the distribution" - in other words, for classification, if you look at the mean of the distribution you will already get a sense of uncertainty because the mean will be far from the 1/0 extremes. Note that the original dropout paper *already showed* that weight averaging closely approximates the mean of several dropout runs - so, by Gal's own logic, it follows that if you want to get a summary of the model uncertainty in the context of classification, can just make predictions the standard way (insofar as weight averaging is actually approximating the mean correctly, which in my personal experience has been the case).

As of now, based on our experiments, our frontrunning strategy for getting a sense of uncertainty is to calibrate the probability outputs; if a classifier's predicted probabilities are calibrated, and the datapoint you are looking at is within your training set distribution (more on this later), that directly tells you how likely the prediction is to be correct. There was a paper at ICML 2017 on using temperature scaling to calibrate the predictions of deep learning models on images; temperature scaling is a variant of a much older technique called Platt Scaling (Platt scaling just fits a logistic regression model to your predictions in order to obtain the calibrated predictions). Isotonic regression is another method people sometimes like for calibration.

The distinction between the two types of uncertainty that were discussed in the first blog post Anshul linked to is interesting. Epistemic uncertainty is about detecting when a datapoint is outside the training distribution, and aleatoric uncertainty is about knowing when you have missing information in your input (eg: recognizing when something in an image is occluded/overexposed/noisy etc). I haven't given much thought to aleatoric uncertainty since it isn't as prominent in the datasets I work with (though I can imagine it's more prominent in other areas of genomics/medicine; "missing data" is pretty common everywhere, but I suspect in our field it's more obvious when information is missing compared to reinforcement learning). Regarding epistemic uncertainty: in Gal's test-time dropout paper, one of their experiments demonstrated that for predicting CO2 levels they had a large uncertainty when the input was far from the training distribution. However, if the goal is to know when you have an input far from the training distribution, one may be better served by attempting to answer that question by directly looking for out-of-distribution samples. Conveniently enough, in ICLR 2017 a paper was published ("A Baseline for Detecting Misclassified and Out-of-Distribution Examples in Neural Networks" by Hendrycks and Gimpel) that showed out-of-distribution examples tend to have a lower maximum softmax probability: https://openreview.net/forum?id=Hkg4TI9xl. In that paper they didn't compare with test-time dropout, but I looked at papers citing it and found one that said, in the context of anomaly detection: "We consider all of the acquisition functions listed in (Gal et al., 2017)...but found that the maximum confidence of the softmax probabilities (i.e., “variation ratio”) acquisition function used by Hendrycks & Gimpel (2016) gave the best performance" (Bayesian Hypernetworks, from MILA: http://bayesiandeeplearning.org/2017/papers/34.pdf; in that paper, they claim that their Bayesian Hypernetworks approach performed better than the baseline from Hendrycks & Gimpel).

I can't recall if there's a section on domain adaptation in the review, but domain adaptation is pretty useful for bringing out-of-distribution samples back into the training distribution. Off the top of my head, I can think of MMD-minimization techniques (https://arxiv.org/pdf/1502.02791.pdf; https://arxiv.org/abs/1610.04181) and gradient reversal layers (https://arxiv.org/abs/1409.7495).

Anyway, my point in summary is that, for classification, simple test-time dropout does not seem that useful compared to just using the predicted probability, and that we should instead approach the problem of uncertainty as a combination of calibration and out-of-distribution sample detection.

**Comment by AvantiShri** [comment]
I'm going to summarize papers to hit as it will help with writing the review:

1) Test-time dropout by Gal and Gharmani, as discussed extensively above. Here is the blog post Anshul linked to by Gal: http://mlg.eng.cam.ac.uk/yarin/blog_3d801aa532c1ce.html

2) The other blog post Anshul linked to which discussed the difference between epistemic and aleatoric uncertainty. Regarding epistemic uncertainty, my recollection is that it mostly just said "we can achieve Bayesian deep learning by putting priors on the weights, but those models are hard to train; an easier thing to do is test-time dropout". https://alexgkendall.com/computer_vision/bayesian_deep_learning_for_safe_ai/

3) Hendrycks & Gimpel on "A Baseline for Detecting Misclassified and Out-of-Distribution Examples in Neural Networks" where they look at the maximum softmax probability. Published in ICLR 2017: https://openreview.net/forum?id=Hkg4TI9xl

4) Citing Hendrycks & Gimpel, there's Bayesian Hypernetworks from MILA (accepted to the Bayesian Deep Learning workshop at NIPS 2017: http://bayesiandeeplearning.org/2017/papers/34.pdf) which had a line suggesting the baseline from Hendrycks & Gimpel outperformed test-time dropout and also claims their technique improves on said baseline.

5) Also citing Hendrycks & Gimpel, there's "Enhancing The Reliability of Out-of-distribution Image Detection in Neural Networks" by Liang et al. (https://arxiv.org/abs/1706.02690; submitted to ICLR 2018: https://openreview.net/forum?id=H1VGkIxRZ). In the abstract it says "Our method is based on the observation that using temperature scaling and adding small perturbations to the input can separate the softmax score distributions between in- and out-of-distribution images...It consistently outperforms the baseline approach by a large margin, establishing a new state-of-the-art performance on this task. For example, ODIN reduces the false positive rate from the baseline 34.7% to 4.3% on the DenseNet (applied to CIFAR-10) when the true positive rate is 95%"

6) Citing both Hendrycks & Gimpel and Liang et al., there's "Training Confidence-calibrated Classifiers for Detecting Out-of-Distribution Samples" (https://arxiv.org/abs/1711.09325). It was also submitted to ICLR 2018 (https://openreview.net/forum?id=ryiAv2xAZ). They modify the loss to encourage the network to make low-confidence predictions for out-of-distribution samples. It was among the accepted abstracts of the Bayesian Deep Learning workshop at NIPS 2017 (http://bayesiandeeplearning.org/2017/papers/24.pdf).

7) Regarding domain adaptation, there's MMD-minimization techniques ("Learning Transferable Features with Deep Adaptation Networks"; https://arxiv.org/pdf/1502.02791.pdf, "Removal of Batch Effects using Distribution-Matching Residual Networks"; https://arxiv.org/abs/1610.04181) and gradient reversal layers ("Unsupervised Domain Adaptation by Backpropagation"; https://arxiv.org/abs/1409.7495)

8) Regarding calibration, there's temperature scaling from ICML 2017 ("On Calibration of Modern Neural Networks"; https://arxiv.org/pdf/1706.04599.pdf) and just plain old Platt Scaling and Isotonic Regression (http://fastml.com/classifier-calibration-with-platts-scaling-and-isotonic-regression/).

Then there's a few other papers I came across that looked relevant:

- "Simple and Scalable Predictive Uncertainty Estimation using Deep Ensembles" (NIPS 2017; http://papers.nips.cc/paper/7219-simple-and-scalable-predictive-uncertainty-estimation-using-deep-ensembles.pdf); claims to perform better than test-time dropout (also known as MC dropout) in the experiments. On a quick glance, they don't seem to have compared with the 'baseline' method from Hendrycks & Gimpel even though they cite it.
- Robustly representing uncertainty through sampling in deep neural networks (among the accepted abstracts of the Bayesian Deep Learning workshop at NIPS 2017: http://bayesiandeeplearning.org/2017/papers/21.pdf). Claims that "Sampling at prediction time increased the calibration of the DNNs’ probabalistic predictions. Sampling weights, whether Gaussian or Bernoulli, led to more robust representation of uncertainty compared to sampling of units. However, using either Gaussian or Bernoulli dropout led to increased test set classification accuracy"). 
- Another paper on hypernetworks at the Bayesian Deep Learning workshop: "Implicit Weight Uncertainty in Neural Networks" (http://bayesiandeeplearning.org/2017/papers/36.pdf). Hypernetworks are networks that model the weights of another network. This paper compares with MC dropout and claims to do better.

**Comment by alexandari** [comment]
I issued a pull request with my input.

**Comment by agitter** [comment]
Closed by #799
